packages/otest/lib/otest.ts

```diff
@@ -0,0 +1,263 @@
+
+${fancy("passing", ansiSequences.greenBg)}: ${result.passedTests.length} ${fancy("failing", ansiSequences.redBg)}: ${result.failingTests.length} ${fancy(
+				"skipped",
+				ansiSequences.yellowBg,
+			)}: ${result.skippedTests.length}`,
+			"\n",
+		)
+
+		for (const test of result.failingTests) {
+			console.error(fancy("FAIL", ansiSequences.redBg), test.path.join(" > "), "|", test.result.name)
+
+			for (const error of test.result.errors) {
+				if (error.userMessage) {
+					console.error(fancy(error.userMessage, ansiSequences.redFg))
+				}
+
+				console.error(error.error)
+				console.log()
+			}
+		}
+	}
+
+	terminateProcess(result: RunResult) {
+		if (typeof process !== "undefined") {
+			process.exit(result.failingTests.length ? 1 : 0)
+		}
+	}
+
+	async runTest(test: Test): Promise<TestResult> {
+		const currentTestResult: TestResult = (this.currentTest = { name: test.name, errors: [], timeout: null, skipped: false })
+
+		let testResolved = false
+
+		async function startTimeoutTask() {
+			await new Promise((resolve) => {
+				if (currentTestResult.timeout == null) throw new Error("timeout not set while running timeout task!")
+				setTimeout(resolve, currentTestResult.timeout)
+			})
+			if (!testResolved) {
+				throw new Error("timed out!")
+			}
+		}
+
+		async function runTask() {
+			try {
+				const p = test.task()
+				currentTestResult.timeout = currentTestResult.timeout ?? OTestmpl.DEFAULT_TIMEOUT_MS
+				await p
+			} finally {
+				testResolved = true
+			}
+		}
+
+		try {
+			// run task and timeout in parallel, if timeout counter comes first, we are timeout out
+			// the test task should set the timeout immediately or we will not pick it up.
+			await Promise.race([runTask(), startTimeoutTask()])
+		} catch (e) {
+			currentTestResult.errors.push({ error: wrapError(e), userMessage: null })
+		} finally {
+			this.currentTest = null
+		}
+		return currentTestResult
+	}
+
+	timeout(ms: number) {
+		if (this.currentTest === null) {
+			throw new Error("timeout() call outside of test")
+		} else if (this.currentTest.timeout != null) {
+			throw new Error(`timeout is already set! ${this.currentTest}`)
+		} else {
+			this.currentTest.timeout = ms
+		}
+	}
+}
+
+function wrapError(e: unknown) {
+	return e instanceof Error ? e : new Error(String(e))
+}
+
+async function promiseMap<T, R>(array: ReadonlyArray<T>, mapper: (value: T) => Promise<R>): Promise<Array<R>> {
+	const result = []
+	for (const el of array) {
+		result.push(await mapper(el))
+	}
+	return result
+}
+
+interface SpecResult {
+	name: string
+	testResults: TestResult[]
+	specResults: SpecResult[]
+}
+
+interface Test {
+	name: string
+
+	task(): Promise<void> | void
+}
+
+interface Spec {
+	name: string
+	tests: Test[]
+	specs: Spec[]
+	before: (() => void)[]
+	after: (() => void)[]
+	beforeEach: (() => void)[]
+	afterEach: (() => void)[]
+}
+
+// this weird API is inherited from ospec, we would like to eventually deprecate it
```

Comment:
replace with what?

Reply:
instead of overloading `o()` invocations we should have `o.test` and `o.check` kind of things